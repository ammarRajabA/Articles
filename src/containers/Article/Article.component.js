import React, {Component} from 'react';
import { connect } from 'react-redux';

import Spinner from '../../components/Spinner/Spinner.component.js';

import history from '../../routes/history';

import {getArticle,deleteArticle} from './Article.actions.js' 

import { Button, Modal, ModalHeader, ModalBody, ModalFooter } from 'reactstrap';

import {stateToHTML} from 'draft-js-export-html';
import {convertFromRaw} from 'draft-js';

import './Article.style.css';

export class Article extends Component{
	state={showDialog:false}
	componentDidMount(){
		if (this.props.match.params.id){
			this.props.getArticle(this.props.match.params.id)
		}
	}
	componentDidUpdate(prevProps){
		if (this.props.match.params.id !== prevProps.match.params.id){
			this.props.getArticle(this.props.match.params.id)
		}
	}
	renderArticle=()=>{
		if (this.props.article.loading) return <Spinner/>
		if (this.props.article.content.length===0) return <Spinner/>
		else{
			var contentState=convertFromRaw(JSON.parse(this.props.article.content))
			return(
					<div className="article">
						<img className="cover" src={this.props.article.image}/>
						<span className="title">{this.props.article.title}</span>
						<span className="author">by {this.props.article.author}</span>
						<div className="content" dangerouslySetInnerHTML={{__html: stateToHTML(contentState)}} />
					</div>
				)
		}
	}
	renderIcons=()=>{
		return (
			<span className="icons-container">
				<span onClick={()=>history.push(`/new-article/${this.props.article.id}`)} className="edit-button">
					<i className="fas fa-edit"></i>
				</span>
				<span onClick={()=>this.setState({showDialog:true})} className="delete-button">
					<i className="fas fa-trash"></i>
				</span>
			</span>
			)
	}
	renderDialog=()=>{
		return (
				<div>
		        		<Modal isOpen={this.state.showDialog} toggle={()=>this.setState({showDialog:false})} className={this.props.className}>
		          		<ModalHeader>Delete Article</ModalHeader>
		          		<ModalBody>
		            		You are about to delete article, are you sure ?
				        </ModalBody>
		          		<ModalFooter>
		            		<Button color="danger" onClick={()=>this.setState({showDialog:false},()=>this.props.deleteArticle(this.props.article.id))}>Yes</Button>{' '}
		            		<Button color="secondary" onClick={()=>this.setState({showDialog:false})}>Cancel</Button>
		          		</ModalFooter>
		        	</Modal>
		      	</div>
			)
	}
	render(){
		return (
				<div className="article-container">
					{this.renderIcons()}
					{this.renderArticle()}
					{this.state.showDialog&&this.renderDialog()}
				</div>
			)
	}
}

const mapStateToProps=(state)=>{
	return {article:state.article}
}

export default connect(mapStateToProps,{getArticle,deleteArticle})(Article);