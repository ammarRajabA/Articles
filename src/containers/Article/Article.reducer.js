const INIT={
	id:"",
	title:"",
	content:"",
	author:"",
	image:"",
	loading:false,
}

export default (state=INIT,action)=>{
	switch(action.type){
		case 'START_GET_ARTICLE':
			return {...state,loading:true}
		case 'FINISH_GET_ARTICLE':
			return {...state,...action.payload,loading:false}
		case 'START_DELETE_ARTICLE':
			return {...state,loading:true}
		case 'FINISH_DELETE_ARTICLE':
			return {...state,loading:false}
		default:
			return state;
	}
}