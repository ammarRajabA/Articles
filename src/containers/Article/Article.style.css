.article-container{
	position: relative;
	padding-bottom: 20px;
}
.article-container .article{
	display: flex;
	flex-direction: column;
	align-items: center;
	margin-top: 20px;
}
.article-container .cover{
	width: 400px;
	height: 300px;
}
.article-container .title{
	font-size: 48px;
	font-weight: bold;
}
.article-container .author{
	font-size: 10px;
	font-style: italic;
}
.article-container .content{
	width: 95%;
}
.article-container .icons-container{
	position: absolute;
	right: 20px;
	top: 10px;
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
}

.article-container .edit-button{
	cursor: pointer;
	font-size: 20px;
	transition: 0.3s;
}
.article-container .edit-button:hover{
	opacity: 0.5;
}
.article-container .delete-button{
	cursor: pointer;
	font-size: 20px;
	transition: 0.3s;
}
.article-container .delete-button:hover{
	opacity: 0.5;
}

@media (min-width: 320px) and (max-width: 767px) {
	.article-container .cover{
		width: 80%;
		height: 300px;
	}
}