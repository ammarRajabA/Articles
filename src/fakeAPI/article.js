import articles from './fake_articles'
export default {
	get:(id)=>{
		return articles.find((article)=>article.id===id)
	},
	delete:(id)=>{
		var target=articles.findIndex((article)=>article.id===id);
		articles.splice(target,1)
	}
}