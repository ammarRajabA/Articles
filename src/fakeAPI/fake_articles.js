export default[
	{
		title:"first",
		content:JSON.stringify({"blocks":[{"key":"5viej","text":"This is my first article, it so cool I can color words ","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":49,"length":5,"style":"color-rgb(243,121,52)"}],"entityRanges":[],"data":{}},{"key":"dh0hv","text":"I can also make bold italic or any other style.","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":16,"length":5,"style":"BOLD"},{"offset":21,"length":7,"style":"ITALIC"}],"entityRanges":[],"data":{}},{"key":"blt1q","text":"I also can write math 25 =32","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":23,"length":2,"style":"SUPERSCRIPT"}],"entityRanges":[],"data":{}},{"key":"ddclt","text":"ðŸ“£ Attention here : My Github account Ammar Rajab ","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":2,"length":48,"style":"color-rgb(209,72,65)"}],"entityRanges":[{"offset":38,"length":11,"key":0}],"data":{}}],"entityMap":{"0":{"type":"LINK","mutability":"MUTABLE","data":{"url":"https://github.com/ammarrajaba","targetOption":"_blank"}}}}),
		image:"https://picsum.photos/id/258/200/300",
		author:"Ammar Rajab",
		id:"13245"
	},
	{
		title:"Second",
		content:JSON.stringify({"blocks":[{"key":"35tp4","text":"Another article by me,","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"a1v72","text":"It shows how we align text","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"cifh6","text":"Left","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"b5gsb","text":"Middle","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{"text-align":"center"}},{"key":"eopjr","text":"and right","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{"text-align":"right"}},{"key":"1h6bs","text":"","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"d3mrc","text":"and change font","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":0,"length":15,"style":"fontfamily-Impact"}],"entityRanges":[],"data":{}}],"entityMap":{}}),
		image:"https://picsum.photos/id/232/200/300",
		author:"Ammar Rajab",
		id:"15555"
	},
	{
		title:"Third",
		content:JSON.stringify({"blocks":[{"key":"5viej","text":"This is my first article, it so cool I can color words ","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":49,"length":5,"style":"color-rgb(243,121,52)"}],"entityRanges":[],"data":{}},{"key":"dh0hv","text":"I can also make bold italic or any other style.","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":16,"length":5,"style":"BOLD"},{"offset":21,"length":7,"style":"ITALIC"}],"entityRanges":[],"data":{}},{"key":"blt1q","text":"I also can write math 25 =32","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":23,"length":2,"style":"SUPERSCRIPT"}],"entityRanges":[],"data":{}},{"key":"ddclt","text":"ðŸ“£ Attention here : My Github account Ammar Rajab ","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":2,"length":48,"style":"color-rgb(209,72,65)"}],"entityRanges":[{"offset":38,"length":11,"key":0}],"data":{}}],"entityMap":{"0":{"type":"LINK","mutability":"MUTABLE","data":{"url":"https://github.com/ammarrajaba","targetOption":"_blank"}}}}),
		image:"https://picsum.photos/id/258/200/300",
		author:"Ammar Rajab",
		id:"13445"
	},
	{
		title:"Forth",
		content:JSON.stringify({"blocks":[{"key":"35tp4","text":"Another article by me,","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"a1v72","text":"It shows how we align text","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"cifh6","text":"Left","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"b5gsb","text":"Middle","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{"text-align":"center"}},{"key":"eopjr","text":"and right","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{"text-align":"right"}},{"key":"1h6bs","text":"","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"d3mrc","text":"and change font","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":0,"length":15,"style":"fontfamily-Impact"}],"entityRanges":[],"data":{}}],"entityMap":{}}),
		image:"https://picsum.photos/id/232/200/300",
		author:"Ammar Rajab",
		id:"15005"
	}
]
