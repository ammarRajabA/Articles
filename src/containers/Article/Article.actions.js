import article from '../../fakeAPI/article'
import {getArticles} from '../Articles/Articles.actions'
import history from '../../routes/history'

export const getArticle=(id)=>dispatch=>{
	dispatch({type:"START_GET_ARTICLE"});
	// Define here Axios or Fetch Http requests
	var fakeResponseArticle=article.get(id);
	console.log(fakeResponseArticle)
	dispatch({type:"FINISH_GET_ARTICLE",payload:fakeResponseArticle});
}

export const deleteArticle=(id)=>dispatch=>{
	dispatch({type:"START_DELETE_ARTICLE"});
	// Define here Axios or Fetch Http requests
	var fakeResponseArticle=article.delete(id);
	dispatch({type:"FINISH_DELETE_ARTICLE",payload:fakeResponseArticle});
	getArticles()(dispatch);
	history.push('/articles');
}