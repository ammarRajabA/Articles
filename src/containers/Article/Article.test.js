import React from 'react'
import { shallow ,mount } from 'enzyme';

import {Article} from './Article.component';
import fake_articles from '../../fakeAPI/fake_articles';

import article from '../../fakeAPI/article'

const test_props={
	match:{
		params:{
			id:"13245"
		}
	},
	article:{
		id:"",
		title:"",
		content:"",
		author:"",
		image:"",
		loading:false,
	}
}

describe('Article component',function () {

	it('match snapshot',function () {
		const component = mount(
			<Article 
				match={test_props.match} 
				article={test_props.article} 
				getArticle={()=>test_props.article=article.get(test_props.match.params.id)}
			/>);
		
		const spy = jest.spyOn(Article.prototype, "componentDidUpdate");		
		expect(spy).toHaveBeenCalledTimes(0);

		component.setProps({...test_props});

		expect(spy).toHaveBeenCalledTimes(1);
		
		expect(component).toMatchSnapshot();
	})
})