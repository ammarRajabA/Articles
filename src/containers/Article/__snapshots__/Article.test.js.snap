// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article component match snapshot 1`] = `
<Article
  article={
    Object {
      "author": "Ammar Rajab",
      "content": "{\\"blocks\\":[{\\"key\\":\\"5viej\\",\\"text\\":\\"This is my first article, it so cool I can color words \\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[{\\"offset\\":49,\\"length\\":5,\\"style\\":\\"color-rgb(243,121,52)\\"}],\\"entityRanges\\":[],\\"data\\":{}},{\\"key\\":\\"dh0hv\\",\\"text\\":\\"I can also make bold italic or any other style.\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[{\\"offset\\":16,\\"length\\":5,\\"style\\":\\"BOLD\\"},{\\"offset\\":21,\\"length\\":7,\\"style\\":\\"ITALIC\\"}],\\"entityRanges\\":[],\\"data\\":{}},{\\"key\\":\\"blt1q\\",\\"text\\":\\"I also can write math 25 =32\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[{\\"offset\\":23,\\"length\\":2,\\"style\\":\\"SUPERSCRIPT\\"}],\\"entityRanges\\":[],\\"data\\":{}},{\\"key\\":\\"ddclt\\",\\"text\\":\\"ðŸ“£ Attention here : My Github account Ammar Rajab \\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[{\\"offset\\":2,\\"length\\":48,\\"style\\":\\"color-rgb(209,72,65)\\"}],\\"entityRanges\\":[{\\"offset\\":38,\\"length\\":11,\\"key\\":0}],\\"data\\":{}}],\\"entityMap\\":{\\"0\\":{\\"type\\":\\"LINK\\",\\"mutability\\":\\"MUTABLE\\",\\"data\\":{\\"url\\":\\"https://github.com/ammarrajaba\\",\\"targetOption\\":\\"_blank\\"}}}}",
      "id": "13245",
      "image": "https://picsum.photos/id/258/200/300",
      "title": "first",
    }
  }
  getArticle={[Function]}
  match={
    Object {
      "params": Object {
        "id": "13245",
      },
    }
  }
>
  <div
    className="article-container"
  >
    <span
      className="icons-container"
    >
      <span
        className="edit-button"
        onClick={[Function]}
      >
        <i
          className="fas fa-edit"
        />
      </span>
      <span
        className="delete-button"
        onClick={[Function]}
      >
        <i
          className="fas fa-trash"
        />
      </span>
    </span>
    <div
      className="article"
    >
      <img
        className="cover"
        src="https://picsum.photos/id/258/200/300"
      />
      <span
        className="title"
      >
        first
      </span>
      <span
        className="author"
      >
        by 
        Ammar Rajab
      </span>
      <div
        className="content"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<p>This is my first article, it so cool I can color words&nbsp;</p>
<p>I can also make <strong>bold </strong><em>italic </em>or any other style.</p>
<p>I also can write math 25 =32</p>
<p>ðŸ“£ Attention here : My Github account A<a href=\\"https://github.com/ammarrajaba\\">mmar Rajab&nbsp;</a></p>",
          }
        }
      />
    </div>
  </div>
</Article>
`;
